wasm/leb128/read.go

```diff
@@ -7,66 +7,81 @@
 package leb128
 
 import (
+	"errors"
 	"io"
 )
 
-// ReadVarUint32 reads a LEB128 encoded unsigned 32-bit integer from r, and
-// returns the integer value, and the error (if any).
-func ReadVarUint32(r io.Reader) (uint32, error) {
-	var (
-		b     = make([]byte, 1)
-		shift uint
-		res   uint32
-		err   error
-	)
-	for {
-		if _, err = io.ReadFull(r, b); err != nil {
-			return res, err
+// read uN defined in https://webassembly.github.io/spec/core/binary/values.html#binary-int
+// note: the result type is uint64, so N must <= 64.
+func readVarUint(r io.Reader, N uint) (uint64, error) {
+	p := make([]byte, 1)
+	_, err := r.Read(p)
```

Comment:
we should probably use `io.ReadFull` instead of just an `io.Reader.Read` to properly handle `wasm` files read over the network (or any non-completely reliable media).